daemon/seahorse-daemon.1
.\"
.\" This page was created on 2006-05-06 19:28:06 by makeman.pl
.\" ``makeman.pl'' is part of the ``MakeMan'' project.
.\" For more information, please see http://mama.sourceforge.net
.\"
.TH seahorse\-daemon 1 "May 04, 2006" "seahorse" "GNOME" 

.SH NAME
seahorse\-daemon \- seahorse pass phrase caching agent

.SH SYNOPSIS
\fBseahorse\-daemon\fR
[ \fIOPTION\fR... ] 

.SH "DESCRIPTION"
.PP
Seahorse is a GNOME application for managing encryption keys. 
.PP
This is the daemon program which provides services to other parts of Seahorse.

.SH "OPTIONS"

.SS "Help options"
.\" Begin List
.TP
\fB\-?, \-\-help\fR
Show summary of all options and exit.
.TP
\fB\-\-usage\fR
Display brief usage message and exit.
.\" End List

.SS "Application options"
.\" Begin List
.TP
\fB\-d, \-\-no\-daemonize\fR
Do not daemonize seahorse\-daemon.
.\" End List

.SS "Additional options"
.PP
In addition, the usual GTK+ and GNOME command line options apply. See the output of \-\-help for details.

.SH "VERSION"
.PP
This man page describes \fBseahorse\-daemon\fR version 0.9.x.

.SH "BUGS"
.PP
If you find a bug, please report it at 
 \fIhttp://bugzilla.gnome.org/enter_bug.cgi?product=seahorse\fR.

.SH "AUTHORS"
.PP
Jacob
Perkins
<jap1@users.sourceforge.net>
.PP
Jose Carlos Garcia
Sogo
<jsogo@users.sourceforge.net>
.PP
Jean
Schurger
<jk24@users.sourceforge.net>
.PP
Stef
Walter
<stef@memberwebs.com>
.PP
Adam
Schreiber
<sadam@clemson.edu>
.PP
This man page was originally written by
Michael
Mende
<debian@menole.net>
for the Debian system.

.SH "COPYRIGHT"
.PP
.br
Copyright (C) 2001, 2002 Jose Carlos Garcia Sogo
.br
Copyright (C) 2002, 2003 Jacob Perkins
.br
Copyright (C) 2004, 2006 Stef Walter, Adam Schreiber

.SH "LICENSE"
.PP
.br
You are free to distribute this software under the terms of the GNU General Public License.
.br
Documentation and help files are covered by GFDL license.

.SH "SEE ALSO"
.PP
\fBseahorse-tool(1)\fR,
\fBseahorse-agent(1)\fR,


NEWS
seahorse 2.31.91
----------------

    * Documentation [Thorsten Sick]
    * Use GDK accessor functions [Adam Schreiber]
    * Use smaller icons in "new" dialog [Adam Schreiber]
    * Build fixes [Adam Schreiber, Pablo Castellano, Matthias Ferdinand]

seahorse 2.30.1
---------------

    * Fixes all but one issue for compiling with -DGSEAL_ENABLE. [Andre Klapper]
    * Remove deprecated GTK+ symbol [Javier Jardón]
    
    Translations:
    * Catalan [Jordi Serratosa]
    * Catalan (Valencian) [Carles Ferrando]
    * Indonesian [Andika Triwidada]
    * Latvian [Peteris Krisjanis]
    * Spanish [Jorge González]


seahorse 2.30.0
---------------

    * Source Documentation [Thorsten Sick]

    Translations:    
    * Bengali [Jamil Ahmed]
    * Basque [Inaki Larranaga Murgoitio]
    * Brazilian Portuguese [Antonio Fernandes C. Neto]
    * British English [Bruce Cowan]
    * Bulgarian [Alexander Shopov]
    * Catalan [Gil Forcada]
    * Czech [Marek Černocký]
    * Danish [Ask H. Larsen]
    * Dutch [Wouter Bolsterlee]
    * Finnish [Tommi Vainikainen]
    * Greek [Nikos Bakaoukas, Simos Xenitellis]
    * Hungarian [Gabor Kelemen]
    * Korean [Changwoo Ryu]
    * Latvian [Peteris Krisjanis]
    * Lithuanian [Gintautas Miliauskas]
    * Polish [Piotr Drąg]
    * Portuguese [Duarte Loreto]
    * Punjabi [A S Alam]
    * Russian [Leonid Kanter]
    * Serbian [Branko Kokanović]
    * Swedish [Daniel Nylander]
    * Ukrainian [Maxim V. Dziumanenko]

seahorse 2.29.91
----------------

    * Change default key lengths for subkey generation. [Adam Schreiber]
    * Remove unused variables [Pablo Castellano]
    * Revoking subkey now works again. Also minor documentation changes. [Pablo Castellano]
    * Don't show the passphrase in plaintext. [Pablo Castellano]
    * Check the OpenPGP engine only [nobled]
    * Fixed wrong variable names in comments [Pablo Castellano]
    * Clean up version constants [nobled]
    * Fixed two warnings at compile time [Pablo Castellano]
    * Unescape URI's before presenting them to the user. [Adam Schreiber]
    * Updated year in the copyright string of the about dialog [Pablo Castellano]
    * Fixed incorrect signal name [Pablo Castellano]
    * Fixed bug in the public key properties GUI. [Pablo Castellano]
    
    Translations:
    * Basque [Inaki Larranaga Murgoitio]
    * Estonian [Ivar Smolin]
    * French [Bruno Brouard]
    * Galician [Fran Diéguez]
    * Italian [Milo Casagrande]
    * Norwegian bokmål [Kjartan Maraas]
    * Romanian [Lucian Adrian Grijincu]
    * Slovenian [Matej Urbančič]
    * Tamil [vasudeven]
    * Traditional Chinese (Hong Kong and Taiwan) [Chao-Hsiung Liao]
    

seahorse 2.29.90
----------------

    * pgp: Make subkey creation more robust [nobled]
    * Drop daemon autostart file [Martin Pitt]
    * Update eggdesktopfile.[ch] from libegg/smclient [Adam Schreiber]    
    * Various fixes [Christian Kirbach]

    Translations:
    * Bengali [Jamil Ahmed]
    * Estonian [Ivar Smolin]
    * German [Christian Kirbach, Mario Blättermann]
    * Slovenian [Matej Urbančič]
    * Spanish [Jorge González]
    * Thai [Theppitak Karoonboonyanan]

seahorse 2.29.4
---------------

    * Experimental support for introspection of libcryptui [Adam Schreiber]
    * Full documentation for seahorse-context.c [Thorsten Sick]
    * Do not use fixed MAXPATHLEN size buffers for portability [Emilio Pozuelo
      Monfort]
    * Add seahorse man page [Pablo Castellano]
    * Fix build with latest gp11 [Vincent Untz]

    Translations:
    * Romanian [Adi Roiban]
    * Chinese [Tao Wei]

seahorse 2.29.3
---------------

    * Set default key length to 2048 bits [Adam Schreiber]
    * Generate RSA-sign/RSA-encrypt key in one pass [nobled]
    * Support GPG 2.0 [Jeff Cai]
    * Fix window signals [Pablo Castellano]
    * GTK-doc comments [Thorsten Sick]
    * SSH Key importing fix [Jeff Cai]

    Translations:
    * Romanian [Adi Roiban]
    * Shavian [Thomas Thurman]
    * Latvian [Peteris Krisjanis]
    * Catalan [Carles Ferrando]
    * Japanese [Takayuki KUSANO]

seahorse 2.29.1
---------------

    * None

seahorse 2.28.1
---------------

    * Prevent opening help window twice. [Pablo Castellano]
    * Set bits setting properly when creating SSH key. [Stef Walter]
    * Add support for GnuPG DSA2 keys. [nobled]
    * Fix subkey creation on newer GnuPG versions [nobled]
    * Fix location label in SSH key properties. [Stef Walter]
    * Fix assertion on exit. [Stef Walter]
    * Fix pasting into filter field. [Adam Schreiber]
    * Set key usage flags properly. [nobled]
    * Fix property bindings. [Bastien Nocera]
    * Internal documentation. [Thorsten Sick]
    * Translation fixes. [Stef Walter]
    * Build fixes. [Stef Walter]

seahorse 2.28.0
---------------

    * void function should not return a value. [Jeff Cai]

seahorse 2.27.92
----------------

    * Include config.h to avoid build breakage when gpgme is 
      compiled with LFS [Vincent Untz]
    * Fix problems with gtk-builder UI files. [Pablo Castellano]
    * Fix problem with 'Type:' label of SSH key displaying 
      location. [Stef Walter]
    * Update context menu display after keyring op. [Stef Walter]
    * Fix more compiler warnings. [Stef Walter, Pablo Castellano]
    * Optionally use silent automake build rules. [Pablo Castellano]
    * Clean up old libglade and CVS files. [Pablo Castellano]
    * Respect the button-images setting, and cleanup gtkbuilder 
      xml. [Matthias Clasen]
    * Add keyring docs. [Paul Cutler]
    * Lots of documentation work. [Adam Schreiber, Thorsten Sick]

seahorse 2.27.90
----------------

    * Fix problem with missing password tab.

seahorse 2.27.5
---------------

    * Fix logic for setting owner trust combo.
    * More Code and API documentation.
    * Old result in search dialog is selected.
    * seahorse-daemon should autostart by default
    * Add a --version parameter to seahorse.
    * No longer daemonizes when autostarted by D-Bus.

daemon/seahorse-daemon.c
/*
 * Seahorse
 *
 * Copyright (C) 2005 Stefan Walter
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the
 * Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "config.h"

#include "seahorse-context.h"
#include "seahorse-daemon.h"
#include "seahorse-gconf.h"
#include "seahorse-gtkstock.h"
#include "seahorse-secure-memory.h"
#include "seahorse-unix-signal.h"

#include "common/seahorse-cleanup.h"
#include "common/seahorse-registry.h"

#ifdef WITH_PGP
#include "pgp/seahorse-pgp-module.h"
#endif

#ifdef WITH_SSH
#include "ssh/seahorse-ssh-module.h"
#endif

#include "libegg/eggsmclient.h"
#include "libegg/eggdesktopfile.h"

#include <glib/gi18n.h>

#include <sys/types.h>
#include <sys/signal.h>

#include <stddef.h>
#include <stdlib.h>
#include <stdio.h>
#include <err.h>
#include <unistd.h>
#include <syslog.h>
#include <fcntl.h>

static gboolean daemon_no_daemonize = FALSE;
static gboolean daemon_running = FALSE;
static gboolean daemon_quit = FALSE;

static const GOptionEntry options[] = {
    { "no-daemonize", 'd', 0, G_OPTION_ARG_NONE, &daemon_no_daemonize, 
        N_("Do not run seahorse-daemon as a daemon"), NULL },
    { NULL }
};

/**
 * SECTION:seahorse-daemon
 * @short_description: The seahorse-daemon. Offering services like DBus, HKP and
 * Avahi
 *
 **/

/**
*
* Daemonizing and some special handling for GPGME
*
**/
static void
daemonize ()
{
    /* 
     * We can't use the normal daemon call, because we have
     * special things to do in the parent after forking 
     */

    pid_t pid;
    int i;

    if (!daemon_no_daemonize) {
        switch ((pid = fork ())) {
        case -1:
            err (1, _("couldn't fork process"));
            break;

        /* The child */
        case 0:
            if (setsid () == -1)
                err (1, _("couldn't create new process group"));

            /* Close std descriptors */
            for (i = 0; i <= 2; i++)
                close (i);
            
            /* Open stdin, stdout and stderr. GPGME doesn't work without this */
            open ("/dev/null", O_RDONLY, 0666);
            open ("/dev/null", O_WRONLY, 0666);
            open ("/dev/null", O_WRONLY, 0666);

            if (chdir ("/tmp") < 0)
                warn ("couldn't change to /tmp directory");
            
            return; /* Child process returns */
        };
    }

    /* Not daemonizing */
    else {
        pid = getpid ();
    }

    /* The parent process or not daemonizing ... */
    if (!daemon_no_daemonize)
        exit (0);
}

/**
*
* The signal handler for termination
*
**/
static void
unix_signal (int signal)
{
    daemon_quit = TRUE;
    if (daemon_running)
        gtk_main_quit ();
}

/**
* smclient: ignored
* data: ignored
*
* The handler for the "quit" signal
*
**/
static void
smclient_quit (EggSMClient *smclient, gpointer data)
{
    daemon_quit = TRUE;
    if (daemon_running)
        gtk_main_quit ();    
}

/**
* log_domain: will be added to the message. Can be NULL
* log_level: the log level
* message: the log message
* user_data: passed to g_log_default_handler, nothing else
*
* Writes log messages
*
**/
static void
log_handler (const gchar *log_domain, GLogLevelFlags log_level, 
             const gchar *message, gpointer user_data)
{
    int level;

    /* Note that crit and err are the other way around in syslog */
        
    switch (G_LOG_LEVEL_MASK & log_level) {
    case G_LOG_LEVEL_ERROR:
        level = LOG_CRIT;
        break;
    case G_LOG_LEVEL_CRITICAL:
        level = LOG_ERR;
        break;
    case G_LOG_LEVEL_WARNING:
        level = LOG_WARNING;
        break;
    case G_LOG_LEVEL_MESSAGE:
        level = LOG_NOTICE;
        break;
    case G_LOG_LEVEL_INFO:
        level = LOG_INFO;
        break;
    case G_LOG_LEVEL_DEBUG:
        level = LOG_DEBUG;
        break;
    default:
        level = LOG_ERR;
        break;
    }
    
    /* Log to syslog first */
    if (log_domain)
        syslog (level, "%s: %s", log_domain, message);
    else
        syslog (level, "%s", message);
 
    /* And then to default handler for aborting and stuff like that */
    g_log_default_handler (log_domain, log_level, message, user_data); 
}

/**
*
* Prepares the logging, sets the log handler to the function "log_handler"
*
**/
static void
prepare_logging ()
{
    GLogLevelFlags flags = G_LOG_FLAG_FATAL | G_LOG_LEVEL_ERROR | 
                G_LOG_LEVEL_CRITICAL | G_LOG_LEVEL_WARNING | 
                G_LOG_LEVEL_MESSAGE | G_LOG_LEVEL_INFO;
                
    openlog ("seahorse-daemon", LOG_PID, LOG_AUTH);
    
    g_log_set_handler (NULL, flags, log_handler, NULL);
    g_log_set_handler ("Glib", flags, log_handler, NULL);
    g_log_set_handler ("Gtk", flags, log_handler, NULL);
    g_log_set_handler ("Gnome", flags, log_handler, NULL);
}


/**
 * main:
 * @argc: default arguments
 * @argv: default arguments
 *
 * Parses the options, daemonizes, starts HKP, DBus and Avahi
 *
 * Returns: 0 on success
 */
int main(int argc, char* argv[])
{
    GOptionContext *octx = NULL;
    GError *error = NULL;

    g_thread_init (NULL);
    
    seahorse_secure_memory_init ();
    
    octx = g_option_context_new ("");
    g_option_context_add_main_entries (octx, options, GETTEXT_PACKAGE);
    g_option_context_add_group (octx, egg_sm_client_get_option_group ());

    if (!gtk_init_with_args (&argc, &argv, _("Encryption Daemon (Seahorse)"), 
                             (GOptionEntry *)options, GETTEXT_PACKAGE, &error)) {
	    g_printerr ("seahorse-daemon: %s\n", error->message);
	    g_error_free (error);
	    exit (1);
    }
     	 
	g_signal_connect (egg_sm_client_get (), "quit", G_CALLBACK (smclient_quit), NULL);
    egg_set_desktop_file(AUTOSTARTDIR "/seahorse-daemon.desktop");
    
    /*
     * All functions after this point have to print messages
     * nicely and not just called exit()
     */
    daemonize ();

    /* Handle some signals */
    seahorse_unix_signal_register (SIGINT, unix_signal);
    seahorse_unix_signal_register (SIGTERM, unix_signal);

    /* Force gconf to reconnect after daemonizing */
    if (!daemon_no_daemonize)
        seahorse_gconf_disconnect ();

    /* We log to the syslog */
    prepare_logging ();

    /* Insert Icons into Stock */
    seahorse_gtkstock_init ();
   
    /* Make the default SeahorseContext */
    seahorse_context_new (SEAHORSE_CONTEXT_APP | SEAHORSE_CONTEXT_DAEMON);

    /* Load the various components */
#ifdef WITH_PGP
    seahorse_pgp_module_init ();
#endif
#ifdef WITH_SSH
    seahorse_ssh_module_init ();
#endif
    
    seahorse_context_refresh_auto (NULL);
    
    /* Initialize the various daemon components */
    seahorse_dbus_server_init ();

    /* Sometimes we've already gotten a quit signal */
    if(!daemon_quit) {
        daemon_running = TRUE;
        gtk_main ();
        g_message ("left gtk_main\n");
    }

    seahorse_dbus_server_cleanup ();

    g_option_context_free (octx);
    seahorse_context_destroy (SCTX_APP ());
    seahorse_cleanup_perform ();

    return 0;
}
